File: leetcode/Algorithms/199.BinaryTreeRightSideView/Solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def __init__(self):
        self.result = []

    def rightSideView(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        self.travel(root, 1)
        return self.result

    def travel(self, tree_node, depth):
        if tree_node is None:
            return

        if depth > len(self.result):
            self.result.append(tree_node.val)

        self.travel(tree_node.right, depth + 1)
        self.travel(tree_node.left, depth + 1)

File: leetcode/Algorithms/199.BinaryTreeRightSideView/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        Queue<TreeNode> queue = new LinkedList<>();
        TreeNode tmp = new TreeNode(0);

        if (root != null) queue.add(root);
        while (queue.size() != 0) {
            int count = queue.size();
            while (count-- > 0) {
                tmp = queue.poll();
                if (tmp.left != null) queue.add(tmp.left);
                if (tmp.right != null) queue.add(tmp.right);
            }
            result.add(tmp.val);
        }
        return result;
    }
}
